File: ios/BetterBasicStrategy/Models/Shoe.swift
import Foundation

struct ShoeState: Codable {
    var cards: [Card]
    var currentIndex: Int
    var penetrationTrigger: Double

    static func create(penetrationTrigger: Double = 0.75) -> ShoeState {
        var deck: [Card] = []
        for _ in 0..<6 {
            for suit in allSuits {
                for rank in allRanks {
                    deck.append(makeCard(suit: suit, rank: rank))
                }
            }
        }
        // Fisher-Yates shuffle
        for i in stride(from: deck.count - 1, through: 1, by: -1) {
            let j = Int.random(in: 0...i)
            deck.swapAt(i, j)
        }
        return ShoeState(cards: deck, currentIndex: 0, penetrationTrigger: penetrationTrigger)
    }

    var needsReshuffle: Bool {
        let pen = cards.isEmpty ? 1.0 : Double(currentIndex) / Double(cards.count)
        return pen >= penetrationTrigger || currentIndex >= cards.count
    }

    var penetration: Double {
        cards.isEmpty ? 0 : Double(currentIndex) / Double(cards.count)
    }

    var remainingCount: Int { cards.count - currentIndex }

    mutating func deal() -> Card? {
        guard currentIndex < cards.count else { return nil }
        defer { currentIndex += 1 }
        return cards[currentIndex]
    }

    func save() {
        if let data = try? JSONEncoder().encode(self) {
            UserDefaults.standard.set(data, forKey: "bbs_shoe")
        }
    }

    static func load(penetrationTrigger: Double = 0.75) -> ShoeState {
        if let data = UserDefaults.standard.data(forKey: "bbs_shoe"),
           let shoe = try? JSONDecoder().decode(ShoeState.self, from: data),
           shoe.cards.count == 312 {
            return shoe
        }
        return create(penetrationTrigger: penetrationTrigger)
    }

    static func clear() {
        UserDefaults.standard.removeObject(forKey: "bbs_shoe")
    }
}

File: expo-app/src/engine/shoe.ts
import type { Card } from '../types'
import { SUITS, RANKS, makeCard } from './card'

export interface ShoeState {
  cards: Card[]
  currentIndex: number
  penetrationTrigger: number
}

export function createShoe(penetrationTrigger = 0.75): ShoeState {
  const deck: Card[] = []
  for (let i = 0; i < 6; i++) {
    for (const suit of SUITS) {
      for (const rank of RANKS) {
        deck.push(makeCard(suit, rank))
      }
    }
  }
  // Fisher-Yates shuffle
  for (let i = deck.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[deck[i], deck[j]] = [deck[j], deck[i]]
  }
  return { cards: deck, currentIndex: 0, penetrationTrigger }
}

export function dealCard(shoe: ShoeState): { card: Card; shoe: ShoeState } | null {
  if (shoe.currentIndex >= shoe.cards.length) return null
  return { card: shoe.cards[shoe.currentIndex], shoe: { ...shoe, currentIndex: shoe.currentIndex + 1 } }
}

export function needsReshuffle(shoe: ShoeState): boolean {
  const pen = shoe.currentIndex / shoe.cards.length
  return pen >= shoe.penetrationTrigger || shoe.currentIndex >= shoe.cards.length
}

export function penetration(shoe: ShoeState): number {
  return shoe.cards.length > 0 ? shoe.currentIndex / shoe.cards.length : 0
}

export function remainingCount(shoe: ShoeState): number {
  return shoe.cards.length - shoe.currentIndex
}
